python/ray/data/dataset_pipeline.py

```diff
@@ -743,6 +752,13 @@ def _optimize_stages(self):
             )
         self._optimized_stages = optimized_stages
 
+    def _peek(self) -> Dataset[T]:
+        if self._first_dataset is None:
+            self._optimize_stages()
+            self._dataset_iter = PipelineExecutor(self)
+            self._first_dataset = next(self._dataset_iter)
+        return self._first_dataset
```

Comment:
It's interesting that `._peek()` has now become the common pipeline execution path, which are slightly odd semantics. 🤔 I'm wondering if it would be cleaner to isolate optimization + execution a bit, maybe something like this:
```python
def _execute(self) -> Iterator[Dataset[T]]:
    self._optimize_stages()
    return PipelineExecutor(self)

def _peek(self) -> Dataset[T]:
    if self._first_dataset is None:
        self._dataset_iter = self._execute()
        self._first_dataset = next(self._dataset_iter)
    return self._first_dataset

def iter_datasets(self) -> Iterator[Dataset[T]]:
    # ...
    if self._dataset_iter is None:
        # No caching from peek, so we directly trigger execution.
        iter = self._execute()
    else:
        # Use cached dataset and iterator from past peek, and unlink the cached attrs
        # to more eagerly release memory.
        iter = itertools.chain([self._first_dataset], self._dataset_iter)
        self._dataset_iter = None
        self._first_dataset = None
    return iter
```
This would have the added benefit of isolating the peek caching to only when the user is actually peeking at the dataset; if they're only going through the `.iter_dataset()` path, `self._dataset_iter` and `self._first_dataset` are never mutated (set or unset).

What do you think?

Reply:
We anyway will need to cache the first dataset, right? If the user calls iter_dataset() and later asks for schema(), we will need to answer it.
It seems separating out a _execute() doesn't save the _peek() from moving the iterator by one step. There is one eager advancement anyways. My understanding is the actually computation is performed in each advancement of the iterator, so this should be a small "waste" (if any, e.g. if the user never uses iter_dataset(), which would be weird)?